include/nbl/video/IDescriptorPool.h

```diff
@@ -29,12 +33,156 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
             uint32_t count;
         };
 
-        explicit IDescriptorPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, uint32_t _maxSets) : IBackendObject(std::move(dev)), m_maxSets(_maxSets) {}
+    private:
+        struct SCombinedImageSampler
+        {
+            core::smart_refctd_ptr<video::IGPUImageView> view;
+            core::smart_refctd_ptr<video::IGPUSampler> sampler;
+        };
+
+        // This construct makes it so that we don't trigger T's constructors and destructors.
+        // TODO(achal): Should we move it to the core namespace?
+        template <typename T>
+        struct alignas(T) StorageTrivializer
+        {
+            uint8_t storage[sizeof(T)];
+        };
+
+    public:
+        explicit IDescriptorPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, const IDescriptorPool::E_CREATE_FLAGS flags, uint32_t _maxSets, const uint32_t poolSizeCount, const IDescriptorPool::SDescriptorPoolSize* poolSizes) : IBackendObject(std::move(dev)), m_maxSets(_maxSets), m_flags(flags)
+        {
+            memset(m_maxDescriptorCount, 0, asset::EDT_COUNT * sizeof(uint32_t));
+
+            for (auto i = 0; i < poolSizeCount; ++i)
+                m_maxDescriptorCount[poolSizes[i].type]++;
+
+            for (auto i = 0; i < asset::EDT_COUNT; ++i)
+            {
+                if (m_maxDescriptorCount[i] > 0)
+                {
+                    if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+                    {
+                        m_generalAllocatorReservedSpace[i] = _NBL_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(alignof(core::smart_refctd_ptr<asset::IDescriptor>), m_maxDescriptorCount[i] * sizeof(void*), 1u), _NBL_SIMD_ALIGNMENT);
+                        m_generalAllocators[i] = core::GeneralpurposeAddressAllocator<uint32_t>(m_generalAllocatorReservedSpace[i], 0u, 0u, 1u, m_maxDescriptorCount[i] * sizeof(void*), 1u);
+                    }
+                    else
+                    {
+                        m_linearAllocators[i] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[i]*sizeof(void*));
```

Comment:
don't have an address space that matches the byte size of the data you're keeping (don't count in byte offsets)

have an address space that matches the number of descriptors you're keeping (count in index offsets)

the reason is that your `operator[]` arithmetic gets simpler and you can re-use the same offset both for API-agnostic descriptor info (lifetime tracking) and API-specific descriptor info (GL multibind params which are kept SoA)

So let alignment be 1, not `alignof(smart_ptr)` and address space size be `m_maxDescriptorCount[i]` not `sizeof(void*)(m_maxDescriptorCount)*m_maxDescriptorCount[i]`

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#issuecomment-1263203930

Reply:
Resolved.